test/unittests/execute_floating_point_test.cpp

```diff
@@ -440,6 +505,100 @@ TYPED_TEST(execute_floating_point_types, neg)
+            const auto expected_positive =
+                (arg == expected_trunc) ? expected_trunc : expected_trunc + TypeParam{1};
+            EXPECT_THAT(exec(arg), Result(expected_positive)) << arg << ": " << expected_positive;
+
+            // For negative values, the ceil() is trunc().
+            EXPECT_THAT(exec(-arg), Result(-expected_trunc)) << -arg << ": " << -expected_trunc;
+        }
+    }
+    ASSERT_EQ(std::fesetround(FE_TONEAREST), 0);
+}
+
+TYPED_TEST(execute_floating_point_types, floor)
+{
+    auto instance = instantiate(parse(this->get_unop_code(Instr::f32_floor)));
+    const auto exec = [&](auto arg) { return execute(*instance, 0, {arg}); };
+
+    ASSERT_EQ(std::fegetround(), FE_TONEAREST);
+    for (const auto rounding_direction : this->all_rounding_directions)
+    {
+        ASSERT_EQ(std::fesetround(rounding_direction), 0);
+        SCOPED_TRACE(rounding_direction);
+
+        // ffloor(+q) = +0  (if 0 < +q < 1)
+        // (also included in positive_trunc_tests, here for explicitness).
+        EXPECT_THAT(exec(FP<TypeParam>::Limits::denorm_min()), Result(TypeParam{0}));
+        EXPECT_THAT(exec(std::nextafter(TypeParam{1}, TypeParam{0})), Result(TypeParam{0}));
+
+        for (const auto& [arg, expected_trunc] :
+            execute_floating_point_types<TypeParam>::positive_trunc_tests)
+        {
+            // For positive values, the floor() is trunc().
+            EXPECT_THAT(exec(arg), Result(expected_trunc)) << arg << ": " << expected_trunc;
+
+            // For negative values, the floor() is trunc() - 1, unless the input is already an
+            // integer.
+            const auto expected_negative =
+                (arg == expected_trunc) ? -expected_trunc : -expected_trunc - TypeParam{1};
+            EXPECT_THAT(exec(-arg), Result(expected_negative)) << -arg << ": " << expected_negative;
+        }
+    }
+    ASSERT_EQ(std::fesetround(FE_TONEAREST), 0);
+}
+
+TYPED_TEST(execute_floating_point_types, trunc)
+{
+    auto instance = instantiate(parse(this->get_unop_code(Instr::f32_trunc)));
+    const auto exec = [&](auto arg) { return execute(*instance, 0, {arg}); };
+
+    ASSERT_EQ(std::fegetround(), FE_TONEAREST);
+    for (const auto rounding_direction : this->all_rounding_directions)
+    {
+        ASSERT_EQ(std::fesetround(rounding_direction), 0);
+
+        // ftrunc(+q) = +0  (if 0 < +q < 1)
+        // (also included in positive_trunc_tests, here for explicitness).
+        EXPECT_THAT(exec(FP<TypeParam>::Limits::denorm_min()), Result(TypeParam{0}));
+        EXPECT_THAT(exec(std::nextafter(TypeParam{1}, TypeParam{0})), Result(TypeParam{0}));
+
+        // ftrunc(-q) = -0  (if -1 < -q < 0)
+        // (also included in positive_trunc_tests, here for explicitness).
+        EXPECT_THAT(exec(std::nextafter(-TypeParam{1}, TypeParam{0})), Result(-TypeParam{0}));
+        EXPECT_THAT(exec(-FP<TypeParam>::Limits::denorm_min()), Result(-TypeParam{0}));
```

Comment:
Potentially all these would be easier to follow by using `nextafter(0, -1)` ?

Reply:
I don't think so. I prefer using named values. And `nextafter(0, -1)` is probably `-0`.